tests/thread/host/spinlock_test_host.cpp

```diff
@@ -0,0 +1,158 @@
+// Copyright (c) Open Enclave SDK contributors.
+// Licensed under the MIT License.
+
+#include <errno.h>
+#include <openenclave/host.h>
+#include <openenclave/internal/error.h>
+#include <openenclave/internal/tests.h>
+#include <signal.h>
+#include <stdio.h>
+#include <cassert>
+#include <chrono>
+#include <cstdio>
+#include <cstdlib>
+#include <cstring>
+#include <thread>
+#include "thread_u.h"
+
+#define ERRNO_TEST_TIMES 10
+#define LARGE_ENOUGH_MEMORY_SIZE 1024 * 1024 * 1024
+#define NUM_SPINLOCK_TEST_THREADS 2
+
+int g_flag_withlock = 0;
+static uint32_t g_lock = 0;
+static int g_error_num = 0;
+
+void* errno_thread1(void* para)
+{
+    int i = 0;
+    int t_errno = 0;
+    void* buffer = NULL;
+    oe_enclave_t* enclave = (oe_enclave_t*)para;
+
+    for (i = 0; i < ERRNO_TEST_TIMES; i++)
+    {
+        if (g_flag_withlock)
+        {
+            enc_spin_lock(enclave, &g_lock);
+        }
+
+        enc_malloc(enclave, &buffer, LARGE_ENOUGH_MEMORY_SIZE, &t_errno);
+
+        if (ENOMEM != t_errno)
+        {
+            printf(
+                "[errno_thread1]Error check errno, expected %d, got %d.\n",
+                ENOMEM,
+                t_errno);
+            g_error_num |= 0x8;
+            if (g_flag_withlock)
+            {
+                enc_spin_unlock(enclave, &g_lock);
+            }
+        }
+
+        if (g_flag_withlock)
+        {
+            enc_spin_unlock(enclave, &g_lock);
+        }
+    }
+
+    return NULL;
+}
+
+void* errno_thread2(void* para)
+{
+    int i = 0;
+    long res = 0;
+    int t_errno = 0;
+    oe_enclave_t* enclave = (oe_enclave_t*)para;
+
+    for (i = 0; i < ERRNO_TEST_TIMES; i++)
+    {
+        if (g_flag_withlock)
+        {
+            enc_spin_lock(enclave, &g_lock);
+        }
+
+        char num_str[] = "123456789987654321123456789987654321";
+        enc_strtol(enclave, &res, num_str, NULL, 10, &t_errno);
+
+        if (ERANGE != t_errno)
+        {
+            printf("[errno_thread2] expected %d, errno %d.\n", ERANGE, t_errno);
+            g_error_num |= 0x40;
+            if (g_flag_withlock)
+            {
+                enc_spin_unlock(enclave, &g_lock);
+            }
+        }
+
+        if (g_flag_withlock)
+        {
+            enc_spin_unlock(enclave, &g_lock);
+        }
+    }
+    return NULL;
+}
+
+void test_errno_multi_threads_sameenclave(oe_enclave_t* enclave)
+{
+    printf("test_errno_multi_threads_sameenclave Starting\n");
+    g_flag_withlock = 1;
```

Comment:
Since errno is thread specific, is the lock and unlock required here?
I am thinking the host can just call enc_malloc on one thread and enc_strtol on another thread and the erronos should be preserved without acquiring a spinlock. Please correct me if I am wrong.

Reply:
The threads do `g_error_num |= 0x80` and `g_error_num | = 0x40`. So locking is needed.

Reply:
Thanks @anakrish. So the lock is needed on the global variable but not the errno. The description suggests you need the lock to protect errno